Core/src/main/java/com/plotsquared/core/plot/Plot.java

```diff
@@ -2295,115 +2296,70 @@ public Set<Plot> getConnectedPlots() {
+        computeDirectMerged(queueCache, frontier, Direction.SOUTH);
+        computeDirectMerged(queueCache, frontier, Direction.WEST);
+        Plot current;
+        while ((current = frontier.poll()) != null) {
+            if (!current.hasOwner() || current.settings == null) {
+                continue;
             }
-            queuecache.add(tmp);
-            frontier.add(tmp);
+            tmpSet.add(current);
+            queueCache.remove(current);
+            addIfIncluded(current, Direction.NORTH, queueCache, tmpSet, frontier);
+            addIfIncluded(current, Direction.EAST, queueCache, tmpSet, frontier);
+            addIfIncluded(current, Direction.SOUTH, queueCache, tmpSet, frontier);
+            addIfIncluded(current, Direction.WEST, queueCache, tmpSet, frontier);
         }
-        if (this.isMerged(Direction.SOUTH)) {
-            tmp = this.area.getPlotAbs(this.id.getRelative(Direction.SOUTH));
+        tmpSet = Set.copyOf(tmpSet);
+        this.connectedCache = tmpSet;
+        return tmpSet;
+    }
+
+    private void computeDirectMerged(Set<Plot> queueCache, Deque<Plot> frontier, Direction direction) {
+        if (this.isMerged(direction)) {
+            Plot tmp = this.area.getPlotAbs(this.id.getRelative(direction));
             assert tmp != null;
-            if (!tmp.isMerged(Direction.NORTH)) {
+            if (!tmp.isMerged(direction.opposite())) {
                 // invalid merge
                 if (tmp.isOwnerAbs(this.getOwnerAbs())) {
-                    tmp.getSettings().setMerged(Direction.NORTH, true);
+                    tmp.getSettings().setMerged(direction.opposite(), true);
                     DBFunc.setMerged(tmp, tmp.getSettings().getMerged());
                 } else {
-                    this.getSettings().setMerged(Direction.SOUTH, false);
+                    this.getSettings().setMerged(direction, false);
                     DBFunc.setMerged(this, this.getSettings().getMerged());
                 }
             }
-            queuecache.add(tmp);
+            queueCache.add(tmp);
             frontier.add(tmp);
         }
-        if (this.isMerged(Direction.WEST)) {
-            tmp = this.area.getPlotAbs(this.id.getRelative(Direction.WEST));
-            if (!tmp.isMerged(Direction.EAST)) {
-                // invalid merge
-                if (tmp.isOwnerAbs(this.getOwnerAbs())) {
-                    tmp.getSettings().setMerged(Direction.EAST, true);
-                    DBFunc.setMerged(tmp, tmp.getSettings().getMerged());
-                } else {
-                    this.getSettings().setMerged(Direction.WEST, false);
-                    DBFunc.setMerged(this, this.getSettings().getMerged());
-                }
-            }
-            queuecache.add(tmp);
+    }
+
+    private void addIfIncluded(
+            Plot current, Direction
+            direction, Set<Plot> queueCache, Set<Plot> tmpSet, Deque<Plot> frontier
+    ) {
```

Comment:
You're not actually checking the merged status of any of the plots here?

Reply:
You're right, that got lost somehow